packages/components/src/Page/Page.tsx

```diff
@@ -59,10 +53,44 @@ export interface PageProps {
   readonly moreActionsMenu?: SectionProps[];
 }
 
+interface PagePropsWithIntro extends BasePageProps {
+  /**
+   * Content of the page. This supports basic markdown node types
+   * such as `_italic_`, `**bold**`, and `[link name](url)`.
+   *
+   * The `externalIntroLinks` prop can only be used if the
+   * `intro` prop is specified.
+   */
+  readonly intro: string;
+
+  /**
+   * Causes any markdown links in the `intro` prop to open in a new
+   * tab, i.e. with `target="_blank"`.
+   *
+   * Defaults to `false`.
+   */
+  readonly externalIntroLinks?: boolean;
+}
+
+interface PagePropsNoIntro extends BasePageProps {
+  // No descriptions are necessary here, since they are already covered
+  // inside of the PagePropsWithIntro interface.
+  readonly intro?: undefined;
+  readonly externalIntroLinks?: never;
+}
+
+/**
+ * We have a discriminated union here to ensure usage
+ * of externalIntroLinks is only valid if intro is
+ * also specified.
+ */
+export type PageProps = PagePropsWithIntro | PagePropsNoIntro;
```

Comment:
I still think the `XOR` is the right choice here.

As noted in the [docs](https://www.npmjs.com/package/ts-xor)

> Typescript's union operator (|) allows combining two types A and B, into a superset type C which can contain all the members of both A and B.

> But sometimes the requirements dictate that we combine two types with mutually exclusive members. So take the members A.a and B.b. Given type C = A | B then we want to impose the restriction that we can set either C.a or C.b but never both AND at least one of the two!

Using a union forces us to redefine `intro` and `externalIntroLinks` as you have done here.

If we are to use `XOR` we don't need to redefine this, because we are saying that `PageProps` will either be `BasePageProps` or `PagePropsWithIntro` but never a combination of both.

```ts
export type PageProps = XOR<BasePageProps, PagePropsWithIntro>
```

Reply:
To add on to this in regards to not wanting add a 3rd party package. 

This is a valid point to worry about, however we already use this package in lots of places throughout Atlantis, so it wouldn't add anything to the payload.

Also to note, that this is only a typing library. When we compile the Atlantis components down to javascript, the output of this file is 0.

Reply:
You make good points in favour of using the `XOR` library, and I'm not necessarily opposed to using it, but I’m also not sure we’re on the same page in terms of how my code works with the discriminated union, and I think it would be good to sync up before we make a decision.

The behaviour offered by the `XOR` library is the exact behaviour I've implemented with the discriminated union. The "discriminated" part of the union is that both parts of the union can never be true at the same time (logical XOR) by virtue of how `PagePropsWithIntro` and `PagePropsNoIntro` are written.

I'll attempt to explain:

`PagePropsWithIntro` has `intro:  string`, and `PagePropsNoIntro` has `intro?: undefined`. `intro` can never be both a `string` and `undefined` at the same time, so `intro` acts as the discriminating member of the union.
- If `intro` is specified as a string, then we can _only_ use the `PagePropsWithIntro` part of the union, so we have the option to also specify the link behaviour because of `externalIntroLinks?: boolean`.
- If we don't specify `intro` (i.e. it's undefined), then we can _only_ use the `PagePropsNoIntro` part of the union, so we _are not allowed_ to set link behaviour because of `externalIntroLinks?: never`.

I don't think this behaviour is clear at first glance (otherwise we wouldn't be having this conversation), and that's not a good thing. I was hoping my code comments would suffice to clarify how this works. Does this explanation make sense? If so, do you still want to go with the `XOR` library? I'm totally OK to do that. It's still nice to know we have a pathway out of a mess if something ever happens to the 3rd-party lib.

Reply:
Yep, it makes sense what you are doing here, and if we didn't already have a pattern to do so I think you made the right choice. Ideally this would be done easier in Typescript. For the time being, let's use `XOR` here so that we keep with the current standards currently set in Atlantis.

You mentioned this

> I don't think this behaviour is clear at first glance

This is really the reason that we choose to use `XOR` currently in Atlantis. It makes doing this quite a bit more readable (especially when we get to more detailed types).

Reply:
That works for me! Glad we're on the same page now. I'll make the change. Thanks for your 👀 on this @eddysims!